### PR TITLE
Adding reverse lookup function to LookupExtractor.

### PR DESCRIPTION
**Pull request by b-slim**
This PR adds to the LookupExtractor `unApply`  function in order to perform reverse lookup.

Part of this PR is to clean most of the unused import.

### CODE REVIEW

**Comment by b-slim** [comment]
@cheddar and @drcrallen please checkout the new API, i want your opinion before i start working on the self cached version of JDBC.

**Comment by nishantmonu51** [comment]
can we clean unused imports in a separate PR ? 
Its good to separate unrelated things in separate PR's helps in backporting things to different branches.

File: processing/src/main/java/io/druid/query/extraction/LookupExtractor.java [hunk 1]
```diff
@@ -39,6 +40,16 @@
...
 
+
+  /**
+   * Provide the reverse mapping from a given value to a list of keys
+   * @param value the value to apply the reverse lookup
+   *              Null and empty are considered to be the same value = nullToEmpty(value)
+   *
+   * @return the list of keys that maps to value or empty list
+   */
+  List<String> unApply(String value);
```
Comment (cheddar) [inline processing/src/main/java/io/druid/query/extraction/LookupExtractor.java]: Can we add the bulk lookup methods as well?

**Comment by cheddar** [comment]
Aside from wanting to add the bulk methods as well, I'm :+1: on this.  The primary change is the `LookupExtractor` and then there's a bunch of changes to the only known current extension.

I agree with separating the cleaning of imports.  We should either do that in its own PR or opportunistically as those classes are touched.

For this PR, I could see an argument for adding the usages of the new interface method as well.  I.e. the broker's optimizations that can be enabled by this.  I can also see an argument for doing that in its own PR.  I'm ok with either of them.

**Comment by b-slim** [comment]
@cheddar i will send two PRs one for making LookupExtractor abstract class and add bulk functions and second for the use case of unApply.

**Comment by b-slim** [comment]
@nishantmonu51 moved out the cleaning import.

**Comment by b-slim** [comment]
@fjy please checkout this

File: extensions/kafka-extraction-namespace/src/test/java/io/druid/query/extraction/namespace/TestKafkaExtractionCluster.java [hunk 2]
```diff
@@ -195,7 +198,7 @@ public void sleep(long ms)
     finally {
       zkClient.close();
     }
-    fnCache.clear();
+    //fnCache.clear();
```
Comment (fjy) [inline extensions/kafka-extraction-namespace/src/test/java/io/druid/query/extraction/namespace/TestKafkaExtractionCluster.java]: can we delete commented out code?

File: extensions/namespace-lookup/src/test/java/io/druid/server/namespace/URIExtractionNamespaceFunctionFactoryTest.java [hunk 3]
```diff
@@ -300,6 +317,23 @@ public void simpleTest() throws IOException, ExecutionException, InterruptedExce
   }
 
   @Test
+  public void testReverseFunction() throws InterruptedException
+  {
+    Assert.assertNull(reverseFnCache.get(namespace.getNamespace()));
+    NamespaceExtractionCacheManagersTest.waitFor(manager.schedule(namespace));
+    Function<String, List<String>> reverseFn = reverseFnCache.get(namespace.getNamespace());
+    while (reverseFn == null) {
+      Thread.sleep(1);
```
Comment (fjy) [inline extensions/namespace-lookup/src/test/java/io/druid/server/namespace/URIExtractionNamespaceFunctionFactoryTest.java]: im really scared by sleep statements such as this. They cause random tests to fail in Travis. Can we use a latch to test instead?

File: processing/src/main/java/io/druid/query/extraction/LookupExtractor.java [hunk 4]
```diff
@@ -40,6 +41,18 @@
...
+   * Provide the reverse mapping from a given value to a list of keys
+   * @param value the value to apply the reverse lookup
+   *              Null and empty are considered to be the same value = nullToEmpty(value)
+   *
+   * @return the list of keys that maps to value or empty list.
+   * Note that for the case of a none existing value in the lookup we have to cases either return an empty list OR list with null element.
+   * returning an empty list implies that user want to ignore such a lookup value.
+   * In the other hand returning a list with the null element implies user want to map the none existing value to the key null.
+   */
+  List<String> unApply(String value);
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/extraction/LookupExtractor.java]: do we have any safety checks built in for the case where the reverse lookup uses a significant amount of memory and will begin to impact performance?

**Comment by fjy** [comment]
@b-slim I have a few comments. Most are minor.

File: extensions/kafka-extraction-namespace/src/test/java/io/druid/query/extraction/namespace/TestKafkaExtractionCluster.java (hunk 2 above)
Comment (b-slim) [inline extensions/kafka-extraction-namespace/src/test/java/io/druid/query/extraction/namespace/TestKafkaExtractionCluster.java]: done

File: processing/src/main/java/io/druid/query/extraction/LookupExtractor.java (hunk 4 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/extraction/LookupExtractor.java]: i am not sure what kind of safety checks, but i would say that it is up to the final user of druid to decide if it make more sense the use `unApply` and save lookup operation at the historical level or let the historical do the job.  So the bottom line the final decision will be made by the user.

File: extensions/namespace-lookup/src/test/java/io/druid/server/namespace/URIExtractionNamespaceFunctionFactoryTest.java (hunk 3 above)
Comment (b-slim) [inline extensions/namespace-lookup/src/test/java/io/druid/server/namespace/URIExtractionNamespaceFunctionFactoryTest.java]: fixed by removing it, i am not sure why the old code has it.

**Comment by b-slim** [comment]
@fjy and @cheddar please check this out.

**Comment by fjy** [comment]
:+1:

**Comment by b-slim** [comment]
@fjy build passed ! can u please pull this

**Comment by drcrallen** [comment]
@b-slim : https://travis-ci.org/druid-io/druid/jobs/97728605

**Comment by b-slim** [comment]
@drcrallen  this should fix it https://github.com/druid-io/druid/pull/2110/files or i can add a while loop as you have done for other tests, let me know what work better.

**Pull request merged**